miniwin/src/d3drm/backends/software/renderer.cpp

```diff
@@ -278,61 +288,101 @@ void Direct3DRMSoftwareRenderer::DrawTriangleProjected(
 	Uint8* pixels = (Uint8*) DDBackBuffer->pixels;
 	int pitch = DDBackBuffer->pitch;
 
+	VertexXY verts[3] = {
+		{p0.x, p0.y, p0.z, p0.w, c0, v0.texCoord.u, v0.texCoord.v},
+		{p1.x, p1.y, p1.z, p1.w, c1, v1.texCoord.u, v1.texCoord.v},
+		{p2.x, p2.y, p2.z, p2.w, c2, v2.texCoord.u, v2.texCoord.v}
+	};
+
+	verts[0].u_over_w = v0.texCoord.u / p0.w;
+	verts[0].v_over_w = v0.texCoord.v / p0.w;
+	verts[0].one_over_w = 1.0f / p0.w;
+
+	verts[1].u_over_w = v1.texCoord.u / p1.w;
+	verts[1].v_over_w = v1.texCoord.v / p1.w;
+	verts[1].one_over_w = 1.0f / p1.w;
+
+	verts[2].u_over_w = v2.texCoord.u / p2.w;
+	verts[2].v_over_w = v2.texCoord.v / p2.w;
+	verts[2].one_over_w = 1.0f / p2.w;
+
+	// Sort verts
+	if (verts[0].y > verts[1].y) {
+		std::swap(verts[0], verts[1]);
+	}
+	if (verts[1].y > verts[2].y) {
+		std::swap(verts[1], verts[2]);
+	}
+	if (verts[0].y > verts[1].y) {
```

Comment:
repeated code

Reply:
This is correct. We're sorting a size 3 array here using a  expanded bubble sort.
I suggest to work this out on paper.